### PR TITLE
[5.0] Eloquent belongsToMany chunk

### PR DESCRIPTION
**Pull request by codelight89**
Fixes this issue #6281

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Tests?

**Comment by codelight89** [comment]
Is it ok? Should I squash commits?

File: tests/Database/DatabaseEloquentBelongsToManyTest.php [hunk 1]
```diff
@@ -50,6 +50,54 @@ public function testModelsAreProperlyHydrated()
...
+		$relation->getRelated()->shouldReceive('newCollection')->andReturnUsing(function($array) { return new Collection($array); });
+
+		$callbackExecutionAssertor = m::mock('StdClass');
+		$callbackExecutionAssertor->shouldReceive('checkModelsCount')->with(2)->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelsCount')->with(1)->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('taylor')->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('dayle')->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('graham')->once();
+		$callbackExecutionAssertor->shouldReceive('checkForeignKey')->with('user_id')->times(3);
+		$callbackExecutionAssertor->shouldReceive('checkOtherKey')->with('role_id')->times(3);
```
Comment (GrahamCampbell) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: Wooooow. That's not a good amount of mocking. Better not to mock this.

File: tests/Database/DatabaseEloquentBelongsToManyTest.php [hunk 2]
```diff
@@ -50,6 +50,54 @@ public function testModelsAreProperlyHydrated()
...
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(1)->once();
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(3)->once();
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(5)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(2)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(4)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(6)->once();
+
+		$relation->chunk(2, function($collection) use ($callbackExecutionAssertor) {
+			$callbackExecutionAssertor->checkModelsCount($collection->count());
+			foreach ($collection as $model) {
```
Comment (GrahamCampbell) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: cs

File: tests/Database/DatabaseEloquentBelongsToManyTest.php [hunk 3]
```diff
@@ -50,6 +50,54 @@ public function testModelsAreProperlyHydrated()
...
+		$callbackExecutionAssertor->shouldReceive('checkOtherKey')->with('role_id')->times(3);
+		$callbackExecutionAssertor->shouldReceive('checkConnectionName')->with('foo.connection')->times(3);
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(1)->once();
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(3)->once();
+		$callbackExecutionAssertor->shouldReceive('checkFirstPivotAttribute')->with(5)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(2)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(4)->once();
+		$callbackExecutionAssertor->shouldReceive('checkSecondPivotAttribute')->with(6)->once();
+
+		$relation->chunk(2, function($collection) use ($callbackExecutionAssertor) {
```
Comment (GrahamCampbell) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: cs

**Comment by codelight89** [comment]
You mean delete getConnectionName, checkModelNameAttribute and checkModelsCount?

File: tests/Database/DatabaseEloquentBelongsToManyTest.php [hunk 4]
```diff
@@ -50,6 +50,54 @@ public function testModelsAreProperlyHydrated()
...
+		$relation->getQuery()->shouldReceive('getModels')->times(3)->andReturn(array($model1,$model2),array($model3),array());
+		$relation->getRelated()->shouldReceive('newCollection')->andReturnUsing(function($array) { return new Collection($array); });
+
+		$callbackExecutionAssertor = m::mock('StdClass');
+		$callbackExecutionAssertor->shouldReceive('checkModelsCount')->with(2)->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelsCount')->with(1)->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('taylor')->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('dayle')->once();
+		$callbackExecutionAssertor->shouldReceive('checkModelNameAttribute')->with('graham')->once();
+		$callbackExecutionAssertor->shouldReceive('checkForeignKey')->with('user_id')->times(3);
```
Comment (codelight89) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: test

Comment (codelight89) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: test

**Comment by codelight89** [comment]
Check it, please.

**Comment by GrahamCampbell** [comment]
Code style/coding standard. You need to put all braces on their own line.

**Comment by codelight89** [comment]
Done

**Comment by GrahamCampbell** [comment]
> Done. And what about this bug in other versions of laravel?

There are no plans to backport this fix to 4.x. If it's merged into 5.0, we always merge 5.0 -> 5.1, so it'll be included in all 5.x releases.

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php [hunk 5]
```diff
@@ -187,6 +187,36 @@ public function paginate($perPage = null, $columns = array('*'))
...
+		while (count($models) > 0)
+		{
+			$this->hydratePivotRelation($models);
+
+			$result = $this->related->newCollection($models);
+
+			// On each chunk result set, we will pass them to the callback and then let the
+			// developer take care of everything within the callback, which allows us to
+			// keep the memory low for spinning through large result sets for working.
+			call_user_func($callback, $result);
```
Comment (JosephSilber) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Shouldn't this check for `false` like [the query builder](https://github.com/laravel/framework/blob/5.0/src/Illuminate/Database/Query/Builder.php#L1482-L1485)?

Comment (codelight89) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: there is no check in [the eloquent builder] (https://github.com/laravel/framework/blob/5.0/src/Illuminate/Database/Eloquent/Builder.php#L204)

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Yes there is. Follow the link Joseph provided.

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Oh, right I see. Maybe the Eloquent builder needs a check then?

Comment (JosephSilber) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: BTW, is there any reason this logic is duplicated all over the place?

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Not sure, would be nice to abstract it. Would a trait do the job?

Comment (JosephSilber) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Any reason the eloquent builder doesn't just call it directly on the base query builder?

Comment (codelight89) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: This chunk functions are equivalent except called get()

Comment (codelight89) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Actually the same method must work in belognsToMany class too. Maybe it is a reason to create ChunkableTrait with this method and use it in Query Builder, Eloquent Builder and BelongsToMany Relation?

Comment (codelight89) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: If you agree I can create new pull request, and this one can be closed. I think test that was created for this pull request can be moved to new pull request.

**Comment by codelight89** [comment]
So you will merge this pull request? Or better to do it as I described above.

**Comment by GrahamCampbell** [comment]
Taylor is the one who decides what gets merged, particularly for sensitive components like the database component.

**Comment by codelight89** [comment]
I think with trait it will be looking something like this https://github.com/codelight89/framework/commit/93d8b5e52302b40671b9a36401d219e363e06555

**Comment by taylorotwell** [comment]
Is it not possible to just add the select fields and then call the usual chunk method on the Eloquent builder? Why duplicate the logic?

**Comment by codelight89** [comment]
@taylorotwell, I don't think it can be solved this way. We need to hydrate pivot relation for every chunk result that we pass to callback.
maybe trait is a solution (as I described above)

**Comment by taylorotwell** [comment]
Yeah you could pass it to the callback, and then call the user supplied callback from that callback. Make sense? :)

**Comment by codelight89** [comment]
@taylorotwell, check it.

**Comment by taylorotwell** [comment]
Is this tested? Do we need an integration test here?

**Comment by codelight89** [comment]
@taylorotwell, I added a small test. Think that's enough.

**Pull request merged**